src/kinship/hierarchy-creator.ts

```diff
@@ -0,0 +1,466 @@
+import * as d3 from 'd3';
+import { DataProvider, Indi, Fam } from '../api';
+import { ChildNodes, TreeNode, LinkType, otherSideLinkType } from './api';
+import { HierarchyFilter } from './hierarchy-filter';
+import { IdGenerator } from '../id-generator';
+import { nonEmpty } from '../utils';
+
+export class HierarchyCreator {
+  static readonly UP_FILTER = HierarchyFilter.allRejecting().modify({
+    indiParents: true,
+    spouseParents: true,
+    indiSiblings: true,
+    spouseSiblings: true,
+  });
+  static readonly DOWN_FILTER = HierarchyFilter.allRejecting().modify({
+    children: true,
+  });
+  static readonly ALL_ACCEPTING_FILTER = HierarchyFilter.allAccepting();
+
+  static createHierarchy(
+    data: DataProvider<Indi, Fam>,
+    startEntryId: EntryId
+  ): Hierarchy {
+    return new HierarchyCreator(data, startEntryId).createHierarchy();
+  }
+
+  readonly startEntryId: EntryId; // Id of entry (indi or fam), which is root of the hierarchy
+  readonly startFamIndi: string | null; // If startEntryId field is a fam id, then startFamIndi field can indicate which spouse in this family is the starting point of the hierarchy
+  readonly queuedNodesById = new Map<string, TreeNode>();
+  readonly idGenerator = new IdGenerator();
+
+  private constructor(
+    readonly data: DataProvider<Indi, Fam>,
+    startEntryId: EntryId
+  ) {
+    [this.startEntryId, this.startFamIndi] = this.expandStartId(startEntryId);
+  }
+
+  // Convert entry id to values of startEntryId and startFamIndi fields
+  private expandStartId(startEntryId: EntryId): [EntryId, string | null] {
+    if (startEntryId.isFam) return [startEntryId, null];
+    const indi = this.data.getIndi(startEntryId.id);
+    if (!indi) throw new Error('Invalid startId');
+    const famsIds = indi.getFamiliesAsSpouse();
+    if (famsIds.length) return [EntryId.fam(famsIds[0]), startEntryId.id];
+    return [startEntryId, null];
+  }
+
+  createHierarchy(): Hierarchy {
+    const upRoot = this.idToNode(this.startEntryId, null, null, false);
+    const downRoot = this.idToNode(this.startEntryId, null, null, false);
+    if (!upRoot || !downRoot) throw new Error('Invalid root node');
+    if (this.startFamIndi) {
+      upRoot!.indi = { id: this.startFamIndi };
+      downRoot!.indi = { id: this.startFamIndi };
+    }
+    const layerEnd = { id: 'LAYER END MARKER' } as TreeNode;
+    let queue = [upRoot, layerEnd];
```

Comment:
I don't understand the idea behind 2 queues in one loop. This looks too complicated.
Would this do the same?
```
const queue = [];
this.fillNodeData(upRoot, HierarchyCreator.UP_FILTER);
this.fillNodeData(downRoot, HierarchyCreator.DOWN_FILTER);
for (const childNode of upRoot.childNodes.getAll()) {
  queue.push(childNode);
}
for (const childNode of downRoot.childNodes.getAll()) {
  queue.push(childNode);
}
while (queue.length) {
  const node = queue.shift()!;
  this.fillNodeData(node, HierarchyCreator.ALL_ACCEPTING_FILTER);
  for (const childNode of node.childNodes.getAll()) {
    queue.push(childNode);
  }
}
```